edge/pkg/devicetwin/dmiserver/server.go

```diff
@@ -0,0 +1,242 @@
+	return &pb.MapperRegisterResponse{
+		DeviceList: deviceList,
+		ModelList:  deviceModelList,
+	}, nil
+}
+
+func (s *server) ReportDeviceStatus(ctx context.Context, in *pb.ReportDeviceStatusRequest) (*pb.ReportDeviceStatusResponse, error) {
+	if !s.limiter.Allow() {
+		return nil, fmt.Errorf("fail to report device status because of too many request: %s", in.DeviceName)
+	}
+
+	for _, twin := range in.ReportedDevice.Twins {
+		propertyType, ok := twin.Reported.Metadata[PropertyType]
+		if !ok {
+			errLog := fmt.Sprintf("fail to get propertyType for property %s of device %s", twin.PropertyName, in.DeviceName)
+			klog.Errorf(errLog)
+			return nil, fmt.Errorf(errLog)
+		}
+		msg, err := CreateMessageTwinUpdate(twin.PropertyName, propertyType, twin.Reported.Value)
+		if err != nil {
+			klog.Errorf("fail to create message data for property %s of device %s with err: %v", twin.PropertyName, in.DeviceName, err)
+			return nil, err
+		}
+		handleDeviceTwin(in.DeviceName, msg)
+	}
+
+	return &pb.ReportDeviceStatusResponse{}, nil
+}
+
+func handleDeviceTwin(deviceName string, payload []byte) {
+	topic := dtcommon.DeviceETPrefix + deviceName + dtcommon.TwinETUpdateSuffix
+	target := modules.TwinGroup
+	resource := base64.URLEncoding.EncodeToString([]byte(topic))
+	// routing key will be $hw.<project_id>.events.user.bus.response.cluster.<cluster_id>.node.<node_id>.<base64_topic>
+	message := beehiveModel.NewMessage("").BuildRouter(modules.BusGroup, modules.UserGroup,
+		resource, messagepkg.OperationResponse).FillBody(string(payload))
+
+	beehiveContext.SendToGroup(target, *message)
+}
+
+// CreateMessageTwinUpdate create twin update message.
+func CreateMessageTwinUpdate(name, valueType, value string) ([]byte, error) {
+	var updateMsg DeviceTwinUpdate
+
+	updateMsg.BaseMessage.Timestamp = getTimestamp()
+	updateMsg.Twin = map[string]*types.MsgTwin{}
+	updateMsg.Twin[name] = &types.MsgTwin{}
+	updateMsg.Twin[name].Actual = &types.TwinValue{Value: &value}
+	updateMsg.Twin[name].Metadata = &types.TypeMetadata{Type: valueType}
+
+	msg, err := json.Marshal(updateMsg)
+	return msg, err
+}
+
+func initSock(sockPath string) error {
+	klog.Infof("init uds socket: %s", sockPath)
+	_, err := os.Stat(sockPath)
+	if err == nil {
+		err = os.Remove(sockPath)
+		if err != nil {
+			return err
+		}
+		return nil
+	} else if os.IsNotExist(err) {
+		return nil
+	} else {
+		return fmt.Errorf("fail to stat uds socket path")
+	}
+}
+
+func StartDMIServer(cache *DMICache) {
+	err := initSock(SockPath)
+	if err != nil {
+		klog.Fatalf("failed to remove uds socket with err: %v", err)
+		return
+	}
+
+	lis, err := net.Listen(deviceconst.UnixNetworkType, SockPath)
+	if err != nil {
+		klog.Errorf("failed to start DMI Server with err: %v", err)
+		return
+	}
+
+	limiter := rate.NewLimiter(rate.Every(Limit*time.Millisecond), Burst)
+
+	s := grpc.NewServer()
+	pb.RegisterDeviceManagerServiceServer(s, &server{
+		limiter:  limiter,
+		dmiCache: cache,
+	})
+	reflection.Register(s)
```

Comment:
is this designed? i guess this is not needed

Reply:
`reflection.Register` can enabled the reflection function of grpc server. see https://github.com/grpc/grpc-go/blob/master/Documentation/server-reflection-tutorial.md